Help me (error at line 98)

### 🔍 Have you checked Spicetify.app page for your issue?

- [X] I have checked the FAQ

### 🔍 Is there already an issue for your problem?

- [X] I have checked older issues, open and closed

### ℹ Environment / Computer Info

```Markdown
I am on Linux Debian 12
```


### 📝 Description

Help I am unable to install spicetify also I am a beginner

shows this
li/master/install.sh | sh
FETCHING Version 2.24.2
DOWNLOADING https://github.com/spicetify/spicetify-cli/releases/download/v2.24.2/spicetify-2.24.2-linux-amd64.tar.gz
######################################################################## 100.0%
EXTRACTING /home/greenblockbuster/.spicetify/spicetify.tar.gz
SETTING EXECUTABLE PERMISSIONS TO /home/greenblockbuster/.spicetify/spicetify
REMOVING /home/greenblockbuster/.spicetify/spicetify.tar.gz
sh: 98: [: .bashrc: unexpected operator
spicetify path already set in /home/greenblockbuster/.bashrc, continuing...

spicetify v2.24.2 was installed successfully to /home/greenblockbuster/.spicetify
Run 'spicetify --help' to get started

### 📸 Screenshots

file:///home/greenblockbuster/Pictures/Screenshot_20231007_161412.xcf

Hello,

I encountered the same error on Pop OS 22.04, which is also based on Debian 12.

> Your `.bashrc` is corrupted. It's not spicetify related

No it's not corrupted and yes it's Spicetify related, because `source ~/.bashrc` throws no error, and line 98 is even empty.

Turns out, the error is referring to **line 98 of `install.sh`**, which happens to contain the following :

```bash
if [ "$1" == ".zshrc" ] && [ ! -z "${ZDOTDIR}" ]; then
```

Where `$1` is `.bashrc`, the mention of that file name in the error.*

After opening the script in JetBrains IntelliJ running ShellCheck, I noticed the following warning :

![](https://github.com/spicetify/spicetify-cli/assets/21284089/aa744495-910a-408c-b4a4-a389f2bd842f)

> In POSIX sh, == in place of = is undefined. See [SC2039](https://github.com/koalaman/shellcheck/wiki/SC2039).

The linked page indicates the following :

> ### Testing Equality
> `==` operator is not supported in POSIX `sh`
> Bash:
> ```Bash
> if [ "$a" == "$b" ]; then
>     echo "equal"
> fi
> ```
> POSIX:
> ```sh
> if [ "$a" = "$b" ]; then
>     echo "equal"
> fi
> ```

Then I noticed this kind of `bash` vs `sh` warning was present all over the file, indicating that it was written to be a `bash` script and not an `sh` one, which is normal of course.

So, editing that like isn't indeed the proper fix, but changing the shebang from `sh` to `bash` is.

Therefore, I created PR #2774.

I wonder though, how could this have gone unnoticed since the very beginning of the project :shrug:

--- a/.gitignore
+++ b/.gitignore
@@ -15,6 +15,4 @@ package.json
 # Logs
 install.log
 
-# VSCode
-.vscode
 pnpm-lock.yaml

--- a/.vscode/extensions.json
+++ b/.vscode/extensions.json
@@ -1,3 +1,3 @@
 {
-	"recommendations": ["timonwong.shellcheck", "biomejs.biome", "golang.go"]
+	"recommendations": ["timonwong.shellcheck", "biomejs.biome", "golang.go", "ms-vscode.powershell"]
 }

--- a/.vscode/extensions.json
+++ b/.vscode/extensions.json
@@ -1,0 +1,3 @@
+{
+	"recommendations": ["timonwong.shellcheck", "biomejs.biome", "golang.go"]
+}

--- a/.vscode/settings.json
+++ b/.vscode/settings.json
@@ -1,0 +1,3 @@
+{
+	"editor.defaultFormatter": "biomejs.biome"
+}

--- a/install.sh
+++ b/install.sh
@@ -1,4 +1,4 @@
-#!/bin/sh
+#!/usr/bin/env sh
 # Copyright 2022 khanhas. GPL license.
 # Edited from project Denoland install script (https://github.com/denoland/deno_install)
 

--- a/install.sh
+++ b/install.sh
@@ -138,7 +138,7 @@ log "spicetify v$tag was installed successfully to $spicetify_install"
 log "Run 'spicetify --help' to get started"
 
 echo "Do you want to install spicetify Marketplace? (Y/n)"
-read -r choice
+read -r choice < /dev/tty
 if [ "$choice" = "N" ] || [ "$choice" = "n" ]; then
     echo "spicetify Marketplace installation aborted"
     exit 0

--- a/install.sh
+++ b/install.sh
@@ -137,7 +137,7 @@ echo
 log "spicetify v$tag was installed successfully to $spicetify_install"
 log "Run 'spicetify --help' to get started"
 
-printf "Do you want to install spicetify Marketplace? (Y/n)"
+echo "Do you want to install spicetify Marketplace? (Y/n)"
 read -r choice
 if [ "$choice" = "N" ] || [ "$choice" = "n" ]; then
     echo "spicetify Marketplace installation aborted"

--- a/install.sh
+++ b/install.sh
@@ -1,56 +1,57 @@
-#!/usr/bin/env sh
+#!/bin/sh
 # Copyright 2022 khanhas. GPL license.
 # Edited from project Denoland install script (https://github.com/denoland/deno_install)
 
 set -e
 
 while getopts ":r" arg; do
-	case "${arg}" in
-		"r") override_root=1 ;;
-	esac
+    case "${arg}" in
+        "r") override_root=1 ;;
+        *) echo "Invalid option -${OPTARG}" >&2; exit 1 ;;
+    esac
 done
 
 is_root() {
-	[ "${EUID:-$(id -u)}" -eq 0 ];
+    [ "$(id -u)" -ne 0 ]
 }
 
-if is_root && [ "${override_root:-0}" -eq 0 ]; then
-	echo "The script was ran as root. Script will now exit"
-	echo "If you did not intend to do this, please run the script without root permissions to avoid issues with Spicetify"
-	echo "You can override this behavior by passing `-r` or `--root` flag to this script"
-	exit
+if ! is_root && [ "${override_root:-0}" -eq 0 ]; then
+    echo "The script was ran as root. Script will now exit"
+    echo "If you did not intend to do this, please run the script without root permissions to avoid issues with Spicetify"
+    echo "You can override this behavior by passing '-r' or '--root' flag to this script"
+    exit
 fi
 
 # wipe existing log
-> install.log
+> install.log :
 
 log() {
-	echo $1
-	echo "["$(date +'%H:%M:%S %Y-%m-%d')"]" $1 >> install.log
+    echo "$1"
+    echo "[$(date +'%H:%M:%S %Y-%m-%d')]" "$1" >> install.log
 }
 
 case $(uname -sm) in
-	"Darwin x86_64") target="darwin-amd64" ;;
-	"Darwin arm64") target="darwin-arm64" ;;
-	"Linux x86_64") target="linux-amd64" ;;
-	"Linux aarch64") target="linux-arm64" ;;
-	*) log "Unsupported platform $(uname -sm). x86_64 and arm64 binaries for Linux and Darwin are available."; exit ;;
+    "Darwin x86_64") target="darwin-amd64" ;;
+    "Darwin arm64") target="darwin-arm64" ;;
+    "Linux x86_64") target="linux-amd64" ;;
+    "Linux aarch64") target="linux-arm64" ;;
+    *) log "Unsupported platform $(uname -sm). x86_64 and arm64 binaries for Linux and Darwin are available."; exit ;;
 esac
 
 # check for dependencies
-command -v curl >/dev/null || { log "curl isn't installed\!" >&2; exit 1; }
-command -v tar >/dev/null || { log "tar isn't installed\!" >&2; exit 1; }
-command -v grep >/dev/null || { log "grep isn't installed\!" >&2; exit 1; }
+command -v curl >/dev/null || { log "curl isn't installed!" >&2; exit 1; }
+command -v tar >/dev/null || { log "tar isn't installed!" >&2; exit 1; }
+command -v grep >/dev/null || { log "grep isn't installed!" >&2; exit 1; }
 
 # download uri
 releases_uri=https://github.com/spicetify/spicetify-cli/releases
 if [ $# -gt 0 ]; then
-	tag=$1
+    tag=$1
 else
-	tag=$(curl -LsH 'Accept: application/json' $releases_uri/latest)
-	tag=${tag%\,\"update_url*}
-	tag=${tag##*tag_name\":\"}
-	tag=${tag%\"}
+    tag=$(curl -LsH 'Accept: application/json' $releases_uri/latest)
+    tag=${tag%\,\"update_url*}
+    tag=${tag##*tag_name\":\"}
+    tag=${tag%\"}
 fi
 
 tag=${tag#v}
@@ -80,7 +81,7 @@ log "REMOVING $tar"
 rm "$tar"
 
 notfound() {
-	cat << EOINFO
+    cat << EOINFO
 Manually add the directory to your \$PATH through your shell profile
 export SPICETIFY_INSTALL="$spicetify_install"
 export PATH="\$PATH:$spicetify_install"
@@ -88,58 +89,59 @@ EOINFO
 }
 
 endswith_newline() {
-    [[ $(tail -c1 "$1" | wc -l) -gt 0 ]]
+    [ "$(od -An -c "$1" | tail -1 | grep -o '.$')" = "\n" ]
 }
 
 check() {
-	local path="export PATH=\$PATH:$spicetify_install"
-	local shellrc=$HOME/$1
-	
-	if [ "$1" == ".zshrc" ] && [ ! -z "${ZDOTDIR}" ]; then
-		shellrc=$ZDOTDIR/$1
-	fi
+    path="export PATH=\$PATH:$spicetify_install"
+    shellrc=$HOME/$1
 
-	# Create shellrc if it doesn't exist
-	if ! [ -f $shellrc ]; then
-		log "CREATING $shellrc"
-		touch $shellrc
-	fi
+    if [ "$1" = ".zshrc" ] && [ -n "${ZDOTDIR}" ]; then
+        shellrc=$ZDOTDIR/$1
+    fi
 
-	# Still checking again, in case touch command failed
-	if [ -f $shellrc ]; then
-		if ! grep -q $spicetify_install $shellrc; then
-			log "APPENDING $spicetify_install to PATH in $shellrc"
-			if ! endswith_newline $shellrc; then
-				echo >> $shellrc
-			fi
-			echo ${2:-$path} >> $shellrc
-			log "Restart your shell to have spicetify in your PATH."
-		else
-			log "spicetify path already set in $shellrc, continuing..."
-		fi
-	else
-		notfound
-	fi
+    # Create shellrc if it doesn't exist
+    if ! [ -f "$shellrc" ]; then
+        log "CREATING $shellrc"
+        touch "$shellrc"
+    fi
+
+    # Still checking again, in case touch command failed
+    if [ -f "$shellrc" ]; then
+        if ! grep -q "$spicetify_install" "$shellrc"; then
+            log "APPENDING $spicetify_install to PATH in $shellrc"
+            if ! endswith_newline "$shellrc"; then
+                echo >> "$shellrc"
+            fi
+            echo "${2:-$path}" >> "$shellrc"
+            log "Restart your shell to have spicetify in your PATH."
+        else
+            log "spicetify path already set in $shellrc, continuing..."
+        fi
+    else
+        notfound
+    fi
 }
 
 case $SHELL in
-	*zsh) check ".zshrc" ;;
-	*bash)
-		[ -f "$HOME/.bashrc" ] && check ".bashrc"
-		[ -f "$HOME/.bash_profile" ] && check ".bash_profile"
-	;;
-	*fish) check ".config/fish/config.fish" "fish_add_path $spicetify_install" ;;
-	*) notfound ;;
+    *zsh) check ".zshrc" ;;
+    *bash)
+        [ -f "$HOME/.bashrc" ] && check ".bashrc"
+        [ -f "$HOME/.bash_profile" ] && check ".bash_profile"
+    ;;
+    *fish) check ".config/fish/config.fish" "fish_add_path $spicetify_install" ;;
+    *) notfound ;;
 esac
 
 echo
 log "spicetify v$tag was installed successfully to $spicetify_install"
 log "Run 'spicetify --help' to get started"
 
-read -p "Do you want to install spicetify Marketplace? (Y/n) " choice
-if [[ $choice == [Nn]* ]]; then
-	echo "spicetify Marketplace installation aborted"
-	exit
+printf "Do you want to install spicetify Marketplace? (Y/n)"
+read -r choice
+if [ "$choice" = "N" ] || [ "$choice" = "n" ]; then
+    echo "spicetify Marketplace installation aborted"
+    exit 0
 fi
 echo "Starting the spicetify Marketplace installation script.."
 curl -fsSL "https://raw.githubusercontent.com/spicetify/spicetify-marketplace/main/resources/install.sh" | sh

